flink-runtime/src/main/java/org/apache/flink/runtime/rpc/akka/AkkaRpcActor.java

```diff
@@ -263,14 +274,41 @@ private void handleRpcInvocation(RpcInvocation rpcInvocation) {
 								if (throwable != null) {
 									promise.failure(throwable);
 								} else {
-									promise.success(value);
+									long resultSize = 0;
+									if (isRemoteSender) {
+										try {
+											resultSize = new SerializedValue(value).getByteArray().length;
+											if (resultSize > maximumFramesize) {
+												promise.failure(new Throwable("The result size exceeds the maximum " +
+													"size, result size is " + resultSize + ", maximum frame size is " +
+													maximumFramesize + " ."));
+											} else {
+												promise.success(value);
+											}
+										} catch (IOException e) {
+											promise.failure(e);
+										}
+									} else {
+										promise.success(value);
+									}
 								}
 							});
 
 						Patterns.pipe(promise.future(), getContext().dispatcher()).to(getSender());
 					} else {
-						// tell the sender the result of the computation
-						getSender().tell(new Status.Success(result), getSelf());
+						if (isRemoteSender) {
+							long resultSize = new SerializedValue(result).getByteArray().length;
+							if (resultSize > maximumFramesize) {
+								getSender().tell(new Status.Failure(new Throwable("The result size exceeds the maximum " +
+									"size, result size is " + resultSize + ", maximum frame size is " + maximumFramesize +
+									" .")), getSelf());
+							} else {
```

Comment:
I think this code fragment could be factored out because we use it in the `CompletableFuture` case as well:

`Either<SerializedValue, AkkaRpcException> serializeResultAndVerifySize(value, maximumFramesize)`